django/template/loader_tags.py

```diff
@@ -171,9 +171,16 @@ def render(self, context):
             # If not, try the cache and get_template().
             template_name = template
             cache = context.render_context.dicts[0].setdefault(self, {})
-            template = cache.get(template_name)
+            try:
+                template = cache.get(template_name)
+            except Exception:
+                template = None
             if template is None:
-                template = context.template.engine.get_template(template_name)
+                if isinstance(template_name, list):
```

Comment:
A few things:

1. `except Exception` is too broad for production code. Maybe you were after `except TypeError` for passing in a list as a dictionary key?
2. The caching won't work for lists of template variables, which will be a performance hit for anyone using this in a loop
3. Did you see the referenced ticket https://code.djangoproject.com/ticket/21065 ? It covers being more consistent with the accepted types, which would imply allowing anything iterable, not just lists.

Probably it's safer to rewrite the code to cast `template_name` to a tuple of strings at the start, which should work as a dictionary key in the cache:

```python
template_name = template
if isinstance(template_name, str):
    template_name = (template_name,)
else:
    template_name = tuple(template_name)
cache = context.render_context.dicts[0].setdefault(self, {})

template = cache.get(template_name)
if template is None:
    template = context.template.engine.select_template(template_name)
    cache[template_name] = template
```

If you do this or similar, could you add test cases that would cover other iterables like tuples and the caching paths?

Reply:
@adamchainz Hey if caching won't work for list, so can I use looping their such that for every template it will check for caching and after that we will try to get template through select_template and check the index of both template in (list or tuple) and which index is lower that template should be render ?  
can is that be a solution ?

Reply:
Casting the list to a tuple, as my code example does, allows it to be a key in the cache. Tuples containing immutable objects are immutable, and therefore hashable: https://docs.python.org/3/glossary.html#term-hashable

Reply:
should I add more test case ?

Reply:
No I think the list and tuple cases are sufficient to show this is working. But we do also need a release note and a documentation update.

Reply:
Ok, I add release note and a documentation for it .

Reply:
@adamchainz I have added the release note and documentation.